### ISSUE TITLE
Failure when Explicit VR contains Implicit VR

### ISSUE THREAD

**Issue by naterichman** [Enet4/dicom-rs#521]
So this is another one with an invalid dicom, which is a case I think should be handled.

This is one where a dicom comes in with ExplicitVRLittleEndian TS and some elements don't actually have VRs.  The way the explicit VR decoder is doing that is by setting the VR to `UN` [here](https://github.com/Enet4/dicom-rs/blob/master/encoding/src/decode/explicit_le.rs#L45) which then causes it to read the two extra reserved bytes which may or may not be there depending on what the actual VR is supposed to be.

I've reproduced this by taking a basic pydicom test file, adding a private tag and manually deleting the VR bytes, so the section of the file looks like this:
```
00000316: 11 00 10 00 4c 4f 04 00                (0011, 0010) LO Length: 4
0000031e: 54 65 73 74                            b'Test'
00000322: 11 00 01 10 02 00 00 00                (0011, 1001) None Length: 2
0000032a: 31 32                                  b'12'
```

 so this section of the file looks like this, which gives an error when I `dicom-dump`

```
➜  ~ $ dicom-dump ~/Downloads/mr_broken.dcm
/home/naterichman/Downloads/mr_broken.dcm:
Could not read data set token

Caused by these errors (recent errors listed first):
  1: Could not read 1585713 value bytes for element tagged (0011,1001)
  2: Could not read value from source at position 480
  3: failed to fill whole buffer
```

Part of the reason I think this should be supported is because pydicom supports it :wink:  The way its done in pydicom is [here](https://github.com/pydicom/pydicom/blob/0beb18676c4df71a308ef41b599812ce71df3cc7/src/pydicom/filereader.py#L150) where they first try to read explicit and then if the resulting VR is not alphabetical, they fall back to implicit.  

I'm not sure how that would work with the design here and I fully understand that this library cannot be as dynamic as python, but I think at least the error should be clearer, i.e. if the VR is not present when reading when reading an explicit DS make that the warning instead of trying to parse the length from the next bytes.

Link to documents, I uploaded both the original and the one I edited to be broken: [onedrive](https://flywheelio-my.sharepoint.com/:f:/g/personal/naterichman_flywheel_io/EpgD64aVK75DpZhePnKU8GUBhdgv7cG50RHOgyavfA27YQ?e=HSTthU)

**Comment by Enet4** [comment]
Thank you for reporting. Can you validate that this is a duplicate of https://github.com/Enet4/dicom-rs/issues/169?

It's a rare scenario, but it's true that the parser currently misbehaves when entering nested sequences with the VR UN.

**Comment by naterichman** [comment]
Yep, it seems to be the same thing from what I can tell without having the original file of the issue author, except this is just an invalid element, not the small exception within sequences.   I'm happy to take a crack at this one, except I don't think your proposed solution would work for this particular case

> From the moment it enters a sequence of VR UN, it would remember to always use implicit VR little endian until it reaches the end of that sequence

since this isn't in a sequence. LMK your thoughts

**Comment by Enet4** [comment]
So I see. This is something else all right!

It should be fine to introduce some simple logic there if it helps read that file to completion.

**Comment by nathannncurtis** [comment]
Hi, I'd be interested in taking a crack at this.

I maintain a DICOM processing tool that handles a lot of non-conformant files from clinical sources (CDs, vendor exports, etc.), and I wrote a parser from scratch specifically because I kept hitting this class of issue. My approach for the mixed VR case is to auto-detect at the boundary between the file meta (group 0x0002, always explicit VR LE per spec) and the dataset: check if the bytes after the first non-meta tag look like a valid two-character VR code and, if not, fall back to implicit VR parsing for the rest of the file.

This handles the case described here, where a file declares Explicit VR Little Endian in the transfer syntax but the dataset is actually implicit. It also handles files that switch mid-stream, which I've seen from a few older scanners.

I'm happy to put together a PR if this approach fits with how you'd want it handled in the parser crate. I’d also want to make sure it doesn't break the strict behavior for well-formed files, so I'd gate it behind a flag or make it a fallback that only kicks in when the standard path fails.

Let me know if this is still something you want addressed and if the approach sounds reasonable.

**Comment by Enet4** [comment]
Thank you for volunteering @nathannncurtis! It makes sense to implement this at `dicom_parser`, though it may take digging through multiple layers: `DataSetReader` (lazy or not) will depend on a single `StatefulDecode` implementation, which in turn depends on one of the primitive decoding types in `dicom_encoding::decode`. Other approaches may be possible, but there should be enough flexibility to incorporate the more flexible header decoding routine in a new type which implements `StatefulDecode`, and configure this according to a new option in `DataSetReaderOptions`.

**Comment by nathannncurtis** [comment]
Thanks for walking through the decoder layers, that made it easy to figure out where everything needed to go.
I've got a semi-working implementation. I'll open a PR soon.
